### PR TITLE
Adds Puzzgrid smite to trap enemy gamers in a fiendishly hard puzzle

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request

https://user-images.githubusercontent.com/35135081/167582441-30162486-44d3-4db3-8961-34193121a76d.mp4

Find the four matching groups from sixteen seemingly random clues

Also updates rust-g

Puzzgrids are config and I already have many puzzles prepared, but their answers will be hidden from the general public :)

Answers are put in the chat

## Why It's Good For The Game

Introduces another fun puzzle similar to the swapping puzzle smite

People will love solving fiendishly hard puzzles to unlock their friends

## Changelog

:cl:
admin: Added a new Puzzgrid smite
/:cl:

### CODE REVIEW

**Comment by Colovorat** [comment]
I'm not sure I understood how the puzzle works from video. I just hope I wouldn't need to solve it

**Comment by Shadowflame909** [comment]
What did mothblocks mean by this

It's in English but it feels like I'm looking at jargon.

**Comment by Mothblocks** [comment]
Find the four matching groups from sixteen seemingly random clues @Colovorat

**Comment by Colovorat** [comment]
> Find the four matching groups from sixteen seemingly random clues @Colovorat

Oh you mean find something common between them?

**Comment by JoshAdamPowell** [comment]
It's like only connect. From the video:
China (South), Could be Black, Could be Red, Caspian - Seas
Celsius, Carbon, periodically, Coding Language, Cosmic Constant - C
Companion to Desist, come to an end, Cancel, Coupled with fire - Cease 
Coupled with saw, catch a glimpse, Catholic ruling body, rods and cones function - See

**Comment by Mothblocks** [comment]
Yeah it tells you the descriptions in the chat when you win/fail, but I didn't capture that in the video

**Comment by MrDoomBringer** [comment]
Feedback should be better for failure, have a message appear in the tgui window rather than chat (where people might not be looking)

File: code/datums/components/puzzgrid.dm [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+
+	/// Callback that will be called when you lose, either through running out of time or running out of lives
+	var/datum/callback/on_fail_callback
+
+	var/time_to_finish
+
+	var/list/all_answers
+	var/list/selected_answers = list()
+	var/list/datum/puzzgrid_group/solved_groups = list()
+	var/lives = 3
```
Comment (SuperNovaa41) [inline code/datums/components/puzzgrid.dm:21]: could use some docs

File: code/datums/components/puzzgrid.dm [hunk 2]
```diff
@@ -0,0 +1,272 @@
...
+		var/line = rustg_file_seek_line(PUZZGRID_CONFIG, line_number)
+		if (!line)
+			continue
+
+		var/line_json_decoded = safe_json_decode(line)
+		if (isnull(line_json_decoded))
+			log_config("Line [line_number + 1] in puzzgrids.txt is not a JSON: [line]")
+			continue
+
+		var/datum/puzzgrid/puzzgrid = new
```
Comment (SuperNovaa41) [inline code/datums/components/puzzgrid.dm:216]: should this be qdel'd if it couldn't be returned

File: code/modules/admin/smites/puzzgrid.dm [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+/obj/structure/puzzgrid_effect
+	anchored = TRUE
+	density = TRUE
+	resistance_flags = INDESTRUCTIBLE
+	icon = 'icons/effects/effects.dmi'
+	icon_state = "shield2"
+
+	var/mob/living/victim
+	var/timer
+	var/gib_on_loss
```
Comment (SuperNovaa41) [inline code/modules/admin/smites/puzzgrid.dm:43]: docs :)

**Comment by calliophis-maculiceps** [comment]
is this a reference to the connecting wall on Only Connect?

**Comment by wesoda25** [comment]
I cant wait to solve fiendishly hard puzzles to save my friends!

**Comment by Mothblocks** [comment]
@calliophis-maculiceps Yes

**Comment by Mothblocks** [comment]
@zxaber I agree but I'm bad at spriting

**Comment by Mothblocks** [comment]
Also docs for non-public stuff isn't really necessary but I will do it anyway just for you

File: code/datums/components/puzzgrid.dm [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+	src.puzzgrid = puzzgrid
+	src.on_victory_callback = on_victory_callback
+	src.on_fail_callback = on_fail_callback
+
+	all_answers = puzzgrid.answers.Copy()
+
+	if (!isnull(timer))
+		addtimer(CALLBACK(src, .proc/out_of_time), timer)
+		time_to_finish = world.time + timer
```
[the remark this answers is not in the document]
Comment (Mothblocks) [reply code/datums/components/puzzgrid.dm:46]: No because this is replicated to clients through UI data

File: code/datums/components/puzzgrid.dm [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+	var/list/answers = list()
+	for (var/datum/puzzgrid_group/puzzgrid_group as anything in puzzgrid.groups)
+		var/list/answers_encoded = list()
+		for (var/answer in puzzgrid_group.answers)
+			answers_encoded += html_encode(answer)
+
+		answers += span_boldnotice("<p>[answers_encoded.Join(", ")]</p>") + span_notice("<p>[html_encode(puzzgrid_group.description)]</p>")
+
+	atom_parent.audible_message(answers.Join("<p>-----</p>"))
```
Comment (AnturK) [inline code/datums/components/puzzgrid.dm:159]: Audible seems like weird choice here

File: code/modules/admin/smites/puzzgrid.dm [hunk 6]
```diff
@@ -0,0 +1,118 @@
+/// Turns the user into a puzzgrid
+/datum/smite/puzzgrid
+	name = "Puzzgrid"
+
+	var/timer
+	var/gib_on_loss
+
+/datum/smite/puzzgrid/configure(client/user)
+	var/timer = input(user, "How long should other people have to solve the grid? 0 gives infinite time.", "Puzzgrid", 0) as num | null
```
Comment (AnturK) [inline code/modules/admin/smites/puzzgrid.dm:9]: Any reason for this to not be tgui input ?

**Review by AnturK — COMMENTED** [review]
Minor stuff, looks good otherwise. But yeah could use some basic distinct sprite.

File: code/datums/components/puzzgrid.dm (hunk 5 above)
Comment (Mothblocks) [reply code/datums/components/puzzgrid.dm:159]: Do we have a good way to just show a message to people around you regardless of being blind/deaf, show_message didn't look like it

File: code/modules/admin/smites/puzzgrid.dm (hunk 6 above)
Comment (Mothblocks) [reply code/modules/admin/smites/puzzgrid.dm:9]: tgui number input doesn't support cancelling IIRC? I could be wrong

File: code/datums/components/puzzgrid.dm (hunk 5 above)
Comment (AnturK) [reply code/datums/components/puzzgrid.dm:159]: not really just direct to_chat

Comment (Mothblocks) [reply code/datums/components/puzzgrid.dm:159]: Shame, oh well I'll just loop viewers or something

**Comment by Mothblocks** [comment]
I finished this during the meeting but forgot to commit it I'll do it tomorrow

**Comment by Mothblocks** [comment]
drafting one last time because i want to make a verb to validate the config

**Pull request merged**